### PR TITLE
wi-fi: add idle timeout

### PR DESCRIPTION
**Pull request by ziemleszcz**
JIRA: DTR-225

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (imx6ull).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: wi-fi/lwip/main.c [hunk 1]
```diff
@@ -67,21 +79,54 @@ static ip_static_addr_t ap_addr = {
...
+
+	memcpy(buf, data, len);
+	buf[len] = '\0';
+
+	timeout = strtol(buf, NULL, 0);
+	if (timeout < 0) {
+		wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_ERR, "can't set Wi-Fi idle timeout (min value is 0)\n");
+		return -1;
+	}
+	else if (timeout > UINT32_MAX) {
```
Comment (agkaminski) [inline wi-fi/lwip/main.c:100]: Hmmm, what if long == int? 
```suggestion
	else if (timeout > INT32_MAX) {
```

File: wi-fi/lwip/main.c [hunk 2]
```diff
@@ -93,94 +138,223 @@ int wifi_ap_set_key(const char *key, size_t len)
...
+
+		mutexLock(wifi_common.lock);
+		condWait(wifi_common.cond, wifi_common.lock, cond_timeout * 1000000ULL);
+		mutexUnlock(wifi_common.lock);
+
+		if (update_idle) {
+			mutexLock(wifi_common.lock);
+			wifi_common.idle_current += 1;
+			mutexUnlock(wifi_common.lock);
+		}
```
Comment (agkaminski) [inline wi-fi/lwip/main.c:208]:
```suggestion

		if (update_idle)
			wifi_common.idle_current += 1;
		mutexUnlock(wifi_common.lock);
```

File: wi-fi/lwip/main.c [hunk 3]
```diff
@@ -93,94 +138,223 @@ int wifi_ap_set_key(const char *key, size_t len)
...
 
-	return 0;
+	if (sys_thread_opt_new("wifi-ap", wifi_ap_thread, NULL, WIFI_THREAD_STACKSZ, WIFI_THREAD_PRIO, &wifi_common.tid) < 0) {
+		mutexUnlock(wifi_common.lock);
+		wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_ERR, "can't create Wi-Fi AP thread\n");
+		return -1;
+	}
+
+	while (wifi_common.flags == 0)
+		condWait(wifi_common.cond, wifi_common.lock, 1000000);
```
Comment (agkaminski) [inline wi-fi/lwip/main.c:330]: Do we need timeout if we don't check for timeout condition?
```suggestion
		condWait(wifi_common.cond, wifi_common.lock, 0);
```

File: wi-fi/lwip/main.c (hunk 1 above)
Comment (ziemleszcz) [reply wi-fi/lwip/main.c:100]: If long == int condition is always false which is no problem. Are you worried about compiler warning?

Comment (agkaminski) [reply wi-fi/lwip/main.c:100]: Right, no problem here, sorry

File: wi-fi/lwip/main.c (hunk 3 above)
Comment (ziemleszcz) [reply wi-fi/lwip/main.c:330]: No, it's remnant of debug code.

**Review by agkaminski — APPROVED** [review]

**Pull request merged**